godot-macros/src/docs/markdown_converter.rs

```diff
@@ -0,0 +1,85 @@
+use markdown::{to_mdast, ParseOptions};
+use std::collections::HashMap;
+
+pub fn convert(md: &str) -> String {
+    // github flavored markdown is what rust uses, i think?
+    let n = to_mdast(md, &ParseOptions::gfm()).unwrap();
+    let definitions = n
+        .children()
+        .unwrap()
+        .iter()
+        .filter_map(|n| match n {
+            Node::Definition(definition) => {
+                Some((definition.identifier.clone(), definition.url.clone()))
+            }
+            _ => None,
+        })
+        .collect::<HashMap<String, String>>();
+
+    n.children()
+        .unwrap()
+        .iter()
+        .filter_map(|n| walk_node(n, &definitions))
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+// will change to &mut String if necessary
+fn walk_node(node: &Node, definitions: &HashMap<String, String>) -> Option<String> {
+    use Node::*;
+    Some(match node {
+        Root(root) => nodes(&root.children, definitions, ""),
+        BlockQuote(quote) => format!("“{}”", nodes(&quote.children, definitions, "\n\n")),
+        List(list) => {
+            let text = nodes(&list.children, definitions, "\n");
+            match list.ordered {
+                true => format!("[ol]\n{text}\n[/ol]"),
+                false => format!("[ul]\n{text}\n[/ul]"),
+            }
+        }
+        Break(_) => "\n".into(),
+        InlineCode(markdown::mdast::InlineCode { value, .. }) => format!("[code]{value}[/code]"),
+        Delete(delete) => format!("[s]{}[/s]", nodes(&delete.children, definitions, "")),
+        Emphasis(emphasis) => format!("[i]{}[/i]", nodes(&emphasis.children, definitions, "")),
+        Image(markdown::mdast::Image { url, .. }) => format!("[img]{url}[/img]",),
+        ImageReference(image) => {
+            format!("[img]{}[/img]", definitions.get(&image.identifier).unwrap())
+        }
+        Link(markdown::mdast::Link { url, children, .. }) => {
+            format!("[url={url}]{}[/url]", nodes(children, definitions, ""))
+        }
+        LinkReference(markdown::mdast::LinkReference {
+            identifier,
+            children,
+            ..
+        }) => format!(
+            "[url={}]{}[/url]",
+            definitions.get(identifier).unwrap(),
+            nodes(children, definitions, "")
+        ),
+        Strong(strong) => format!("[b]{}[/b]", nodes(&strong.children, definitions, "")),
+        Text(text) => text.value.clone(),
+        // TODO: more langs?
+        Code(markdown::mdast::Code { value, .. }) => format!("[codeblock]\n{value}\n[/codeblock]"),
+        ListItem(list_item) => nodes(&list_item.children, definitions, ""),
+        Paragraph(paragraph) => nodes(&paragraph.children, definitions, ""),
+        _ => return None,
+    })
+}
+
+fn nodes(nodes: &[Node], definitions: &HashMap<String, String>, separator: &str) -> String {
+    nodes
+        .iter()
+        .filter_map(|n| walk_node(n, definitions))
+        .collect::<Vec<_>>()
+        .join(separator)
+}
```

Comment:
Can you maybe add a short doc to this function, and name it more descriptively? E.g. include a verb describing what you do with the nodes.